article/app/topmentions/TopMentionsS3Client.scala

```diff
@@ -0,0 +1,82 @@
+package topmentions
+
+import com.amazonaws.services.s3.model.{GetObjectRequest, S3Object}
+import com.amazonaws.services.s3.{AmazonS3, AmazonS3Client}
+import com.amazonaws.util.IOUtils
+import common.GuLogging
+import conf.Configuration
+import play.api.libs.json.{JsError, JsSuccess, Json}
+
+import scala.collection.JavaConverters._
+import scala.concurrent.Future
+import scala.util.{Failure, Success, Try}
+import TopMentionsResponse._
+
+trait TopMentionsS3Client {
+  def getListOfKeys(): Future[List[String]]
+  def getObject(key: String): Future[TopMentionsDetails]
+}
+
+final class TopMentionsS3ClientImpl extends TopMentionsS3Client with GuLogging {
+  lazy val bucket = Configuration.aws.topMentionsStoreBucket
+
+  lazy val client: AmazonS3 =
+    AmazonS3Client.builder
+      .withCredentials(Configuration.aws.mandatoryCredentials)
+      .withRegion(conf.Configuration.aws.region)
+      .build()
+
+  def getListOfKeys(): Future[List[String]] = {
+    Try {
+      val s3ObjectList = client.listObjects(bucket).getObjectSummaries().asScala.toList
+      s3ObjectList.map(_.getKey)
+    } match {
+      case Success(value) =>
+        log.info(s"got list of ${value.length} top mentions from S3")
+        Future.successful(value)
+      case Failure(exception) =>
+        log.error(s"failed in getting the list of top mentions from S3 - ${exception.getMessage}")
+        Future.failed(exception)
+    }
+  }
+
+  def getObject(key: String): Future[TopMentionsDetails] = {
+    Try {
+      val request = new GetObjectRequest(bucket, key)
+      val result = client.getObject(request)
+      parse(result)
+    } match {
+      case Success(value) =>
+        log.info(s"got topMentionResponse from S3 for key ${key}")
+        Future.successful(value)
+      case Failure(exception) =>
+        log.error(s"S3 retrieval failed for key ${key} - ${exception.getMessage}")
+        Future.failed(exception)
+    }
+  }
+
+  private def parse(s3Object: S3Object) = {
```

Comment:
As this is a pure function, could it go in a companion object? 
Also, maybe return type definition here would be good so it's clearer what the S3Object is being turned into

Reply:
Great suggestion, I moved the parse method into a companion object and it's now an extension method on S3Object